### ISSUE TITLE
About Hot swap of mmaped Tar

### ISSUE THREAD

**Issue by Guo-astro** [valhalla/valhalla#2156]
Happy new year! 
Wondering if we can load a different set of mapped tar (same build version) without shutting down the server?

**Comment by kevinkreiser** [comment]
Its been discussed in a couple different issues thusfar. At the moment we do not support this. The main issue would be having a method to synchronize the use of the memory mapped pointer with those pointers into it, ie from graphreader et al. What you would have to do is, when you map a different file you must also make sure all graphreaders call `clear` and no operations are holding on to any pointers into the previously mapped memory.

One way to do this would be to have the loki_worker or thor_worker reload their graph readers. That would cause whatever file is now there or whatever file the config now points to, to get memory mapped etc.

The problem is these two worker objects are not synchronized. But for them to work together properly, both need to have the same tile set loaded. because there is a queue between loki and thor workers you can never be sure when things are happening in time. So the only wait to force this synchronization would be to actually have a piece of info in the request that says what graphreader config was used at the beginning of this request and, once the request gets to thor, reload the graph reader if the request is using a different one. presumably there wouldnt be too much loading and unloading of tilesets unless there is a massive backlog of requests in thors queue.

i suppose it would be possible to keep more than one file mapped and switch between them until the older file isnt seeing any more requests.. maybe.. in any case its safe to say its a bit complicated

**Comment by Guo-astro** [comment]
Thanks for the explanation!
> What you would have to do is, when you map a different file you must also make sure all graphreaders call clear and no operations are holding on to any pointers into the previously mapped memory.

Does the clear mean the function   void SimpleTileCache::Clear() ?

> The problem is these two worker objects are not synchronized. But for them to work together properly, both need to have the same tile set loaded. because there is a queue between loki and thor workers you can never be sure when things are happening in time

Where is the queue?

**Comment by kevinkreiser** [comment]
yep `GraphReader::Clear()` and the queue lives inside of `prime_server::proxy_t` which is a process/thread that runs between the pools of loki and thor workers

**Comment by Guo-astro** [comment]
@kevinkreiser  Thanks for the explanation! Another question is is the .gph file format support hot-swapping? I mean once the tile is used, we unload it from memory

**Comment by kevinkreiser** [comment]
yeah so there are multiple modes of workign with tiles. if you are using memory mapping the OS decides when to unmap a region of the tar for certain gph files. if you are not using memory mapping, then we have a mechanism for keeping the amount of memory below a certain threshold via `GraphReader::Clear()` which will remove tiles from the memory cache

**Comment by Guo-astro** [comment]
good！
 so can i folllow the same logic for unloading and loading the mmap tar？
Another question is when using tiles there will be a performance impact when calculating large amount of route （i.e.，source_to_target）because mappping and unmmaping tiles from memory use disk io is that correct？ Is there a benchmark for it？
Thx！


iPhoneから送信
